File: flutter/packages/hatake_core/lib/src/definition/field_definition.dart
import 'package:equatable/equatable.dart';

import 'column_definition.dart';
import 'field_types.dart';
import 'option_item.dart';
import 'options_source.dart';
import 'sub_table_source.dart';
import 'validator_definition.dart';

/// A single input field within a form.
class FieldDefinition extends Equatable {
  /// The backing data key (matches keys in a record `Map`).
  final String field;

  /// Display label.
  final String label;

  /// Field type identifier (see [FieldTypes]). Open string, plugin-extensible.
  final String type;

  /// Whether the field is required. Sugar for a `required` validator; renderers
  /// may also use it to show a required marker.
  final bool required;

  /// Whether the field is read-only.
  final bool readOnly;

  /// Default value applied when creating a new record.
  final Object? defaultValue;

  /// Validation rules for this field.
  final List<ValidatorDefinition> validators;

  /// Options for select / radio / multiSelect field types.
  final List<OptionItem> options;

  /// Parent field whose value narrows [options] (a static cascade).
  ///
  /// The choices offered are those whose `when` equals the parent's current
  /// value, plus any option without a `when`. Null = every option is offered.
  final String? optionsFrom;

  /// Where to fetch the options from, instead of listing them (see
  /// [OptionsSource]). Null = use [options].
  final OptionsSource? optionsSource;

  /// Optional display formatter name (see `FormatterRegistry`). Options are
  /// read from [config].
  final String? format;

  /// Input converters/normalizers applied before validation/persistence
  /// (see `ConverterRegistry`), e.g. `[toHankaku, trim]`.
  final List<String> normalize;

  /// Plugin / renderer specific extra configuration. Kept open so field
  /// plugins can carry arbitrary settings without changing the model.
  final Map<String, Object?> config;

  /// Show this field only when the condition matches the current record
  /// (structured map, see `evaluateCondition`). Null = always visible.
  final Map<String, Object?>? visibleWhen;

  /// Enable this field only when the condition matches. Null = always enabled.
  final Map<String, Object?>? enabledWhen;

  /// Derive this field's value from the record (structured map, see
  /// `ComputedRegistry`). Computed fields are shown read-only.
  final Map<String, Object?>? computed;

  /// Roles allowed to see this field (see `isAllowed`). Empty = everyone.
  /// UI-level display gating only — not access enforcement.
  final List<String> roles;

  /// Child-row grid columns, for `type: subTable` (master-detail). The field's
  /// value is then a list of records, one per row. DSL key: `columns`.
  final List<ColumnDefinition> columns;

  /// Editor fields for one child row, for `type: subTable`. When empty the
  /// renderer derives inputs from [columns]. DSL key: `fields`.
  final List<FieldDefinition> rowFields;

  /// Where child rows come from, for `type: subTable`. Null (the default) keeps
  /// them embedded in the parent record; set it to page them from their own
  /// repository instead. DSL key: `source`.
  final SubTableSource? source;

  const FieldDefinition({
    required this.field,
    required this.label,
    this.type = FieldTypes.text,
    this.required = false,
    this.readOnly = false,
    this.defaultValue,
    this.validators = const [],
    this.options = const [],
    this.optionsFrom,
    this.optionsSource,
    this.format,
    this.normalize = const [],
    this.config = const {},
    this.visibleWhen,
    this.enabledWhen,
    this.computed,
    this.roles = const [],
    this.columns = const [],
    this.rowFields = const [],
    this.source,
  });

  @override
  List<Object?> get props => [
        field,
        label,
        type,
        required,
        readOnly,
        defaultValue,
        validators,
        options,
        optionsFrom,
        optionsSource,
        format,
        normalize,
        config,
        visibleWhen,
        enabledWhen,
        computed,
        roles,
        columns,
        rowFields,
        source,
      ];
}

File: java/src/main/java/io/hatake/core/FieldDefinition.java
package io.hatake.core;

import java.util.List;
import java.util.Map;

/**
 * フォームの入力項目 1 件。
 *
 * <p>{@code requiredWhen} は条件付き必須。表示に関する条件（{@code visibleWhen} /
 * {@code enabledWhen} / {@code readOnlyWhen}）と違って**サーバ側でも効く**
 * （{@link FormValidator} が同じ条件を評価する）。
 *
 * <p>{@code type} が {@link #SUB_TABLE} のときは明細（親子）項目で、値は行レコードの
 * リストになる。{@code columns} は子グリッドの表示形状（DSL キー {@code columns}）、
 * {@code rowFields} は 1 行分の入力項目（DSL キー {@code fields}）。
 * {@code source} があれば子行は親レコードではなく別 Repository にある
 * （{@link SubTableSource}）。
 */
public record FieldDefinition(
        String field,
        String label,
        String type,
        boolean required,
        Map<String, Object> requiredWhen,
        boolean readOnly,
        Map<String, Object> readOnlyWhen,
        List<ValidatorDefinition> validators,
        String format,
        List<String> normalize,
        Map<String, Object> visibleWhen,
        Map<String, Object> enabledWhen,
        Map<String, Object> computed,
        List<String> roles,
        List<ColumnDefinition> columns,
        List<FieldDefinition> rowFields,
        SubTableSource source) {

    /** 明細（master-detail）項目の type。 */
    public static final String SUB_TABLE = "subTable";

    /**
     * 明細に関係しない普通の項目を組み立てる短縮コンストラクタ。
     * 条件と明細用の枠は空で埋める。
     *
     * <p>正式コンストラクタは項目が増えるたびに全呼び出し元を壊すので、
     * 明細を使わない箇所（テスト・単純な組み立て）はこちらを使う。
     */
    public FieldDefinition(
            String field,
            String label,
            String type,
            boolean required,
            boolean readOnly,
            List<ValidatorDefinition> validators,
            String format,
            List<String> normalize) {
        this(field, label, type, required, null, readOnly, null, validators, format,
                normalize, null, null, null, List.of(), List.of(), List.of(), null);
    }

    /** この項目が明細かどうか。 */
    public boolean isSubTable() {
        return SUB_TABLE.equals(type);
    }

    /** 子行が別 Repository にある明細かどうか。 */
    public boolean hasSubTableSource() {
        return source != null;
    }
}
